File: ruby/lib/vedadb/interceptor.rb
# frozen_string_literal: true

module VedaDB
  # Middleware / interceptor pipeline for VedaDB operations.
  #
  # Each interceptor implements `#call(env, &block)` where `env` is a Hash
  # with keys like :operation, :sql, :client, :start_time, etc.
  #
  # Built-in interceptors:
  #   VedaDB::Interceptor::Logging   -> logs every query
  #   VedaDB::Interceptor::Timing    -> records timing metrics
  #   VedaDB::Interceptor::Retry     -> adds retry logic
  #   VedaDB::Interceptor::Circuit   -> wraps with circuit breaker
  #   VedaDB::Interceptor::Cache     -> query-level caching
  #   VedaDB::Interceptor::Validate  -> validates SQL before sending
  #
  # Usage:
  #   client = VedaDB::Client.new("localhost", 6380)
  #   client.use VedaDB::Interceptor::Logging.new(logger: Logger.new($stdout))
  #   client.use VedaDB::Interceptor::Timing.new(metrics: my_metrics)
  class Interceptor
    attr_reader :interceptors

    def initialize
      @interceptors = []
    end

    # Register an interceptor.  Interceptors are invoked in registration order.
    def use(interceptor)
      @interceptors << interceptor
      self
    end

    # Remove all interceptors.
    def clear
      @interceptors.clear
      self
    end

    # Execute a block through the full interceptor chain.
    #
    # @param env [Hash] operation context
    # @yield the actual operation
    # @return the block's return value
    def run(env = {})
      chain = build_chain(@interceptors.dup, proc { yield })
      chain.call(env)
    end

    private

    def build_chain(interceptors, final)
      interceptors.reverse.reduce(final) do |nxt, interceptor|
        proc { |env| interceptor.call(env, &nxt) }
      end
    end

    # ------------------------------------------------------------------
    # Built-in interceptors
    # ------------------------------------------------------------------

    # Log every query at INFO level.
    class Logging
      def initialize(logger: nil, level: :info)
        @logger = logger || default_logger
        @level  = level
      end

      def call(env)
        sql = env[:sql] || env[:query]
        @logger.public_send(@level) { "[VedaDB] #{env[:operation]} | #{sql}" }
        yield env
      rescue StandardError => e
        @logger.error { "[VedaDB] ERROR #{env[:operation]}: #{e.class}: #{e.message}" }
        raise
      end

      private

      def default_logger
        require "logger"
        Logger.new($stdout)
      end
    end

    # Record timing for every operation.
    class Timing
      def initialize(metrics: nil, callback: nil)
        @metrics  = metrics
        @callback = callback
      end

      def call(env)
        start = Process.clock_gettime(Process::CLOCK_MONOTONIC)
        result = yield env
        duration = Process.clock_gettime(Process::CLOCK_MONOTONIC) - start

        env[:duration] = duration
        @metrics&.query_executed(duration: duration, success: true)
        @callback&.call(env.merge(success: true))

        result
      rescue StandardError => e
        duration = Process.clock_gettime(Process::CLOCK_MONOTONIC) - start if start
        env[:duration] = duration
        @metrics&.query_executed(duration: duration || 0, success: false)
        @callback&.call(env.merge(success: false, error: e))
        raise
      end
    end

    # Retry failed operations.
    class Retry
      def initialize(max: 3, base: 0.1, on: [ConnectionError])
        @retry = VedaDB::Retry.new(max: max, base: base, on: on)
      end

      def call(env)
        @retry.call { yield env }
      end
    end

    # Circuit breaker wrapper.
    class Circuit
      def initialize(breaker: nil)
        @breaker = breaker || VedaDB::CircuitBreaker.new
      end

      def call(env)
        @breaker.call { yield env }
      end
    end

    # Query-level caching interceptor.
    class Cache
      def initialize(cache_store, ttl: 60)
        @cache = cache_store
        @ttl   = ttl
      end

      def call(env)
        sql = env[:sql] || env[:query]

        unless sql && sql.match?(/^\s*SELECT/i)
          return yield env
        end

        key = "vedadb:query:#{Digest::SHA256.hexdigest(sql)}"
        cached = @cache.get(key)
        return cached if cached

        result = yield env
        @cache.set(key, result, @ttl)
        result
      end
    end

    # Validate SQL before sending.
    class Validate
      def initialize
        @forbidden = %w[DROP DATABASE SHUTDOWN EXEC xp_ sp_]
      end

      def call(env)
        sql = env[:sql] || env[:query]

        if sql
          @forbidden.each do |pattern|
            if sql.upcase.include?(pattern)
              raise QueryError, "Forbidden SQL pattern detected: #{pattern}"
            end
          end
        end

        yield env
      end
    end

    # Metrics collection interceptor.
    class MetricsCollector
      def initialize(metrics)
        @metrics = metrics
      end

      def call(env)
        @metrics.connection_event(event: "open") if env[:operation] == :connect

        start = Process.clock_gettime(Process::CLOCK_MONOTONIC)
        result = yield env
        duration = Process.clock_gettime(Process::CLOCK_MONOTONIC) - start

        @metrics.query_executed(duration: duration, success: true)
        result
      rescue StandardError
        @metrics.query_executed(duration: 0, success: false)
        raise
      end
    end
  end
end

File: go/vedadb/interceptor.go
package vedadb

import (
	"context"
	"fmt"
	"sync"
	"time"
)

// ---------------------------------------------------------------------------
// Connection Middleware / Interceptor
// ---------------------------------------------------------------------------

// Interceptor is a function that wraps query execution.
// It receives the SQL, args, and a handler function to execute the actual query.
type Interceptor func(ctx context.Context, sql string, args []interface{}, next Handler) (*Result, error)

// Handler is the function signature for the actual query handler.
type Handler func(ctx context.Context, sql string, args []interface{}) (*Result, error)

// InterceptorChain manages a chain of interceptors.
type InterceptorChain struct {
	interceptors []Interceptor
	mu           sync.RWMutex
}

// NewInterceptorChain creates a new interceptor chain.
func NewInterceptorChain() *InterceptorChain {
	return &InterceptorChain{
		interceptors: make([]Interceptor, 0),
	}
}

// Add appends an interceptor to the chain.
func (ic *InterceptorChain) Add(interceptor Interceptor) {
	ic.mu.Lock()
	defer ic.mu.Unlock()
	ic.interceptors = append(ic.interceptors, interceptor)
}

// AddFirst prepends an interceptor to the chain.
func (ic *InterceptorChain) AddFirst(interceptor Interceptor) {
	ic.mu.Lock()
	defer ic.mu.Unlock()
	ic.interceptors = append([]Interceptor{interceptor}, ic.interceptors...)
}

// Remove removes an interceptor by index.
func (ic *InterceptorChain) Remove(index int) error {
	ic.mu.Lock()
	defer ic.mu.Unlock()
	if index < 0 || index >= len(ic.interceptors) {
		return fmt.Errorf("interceptor index %d out of range", index)
	}
	ic.interceptors = append(ic.interceptors[:index], ic.interceptors[index+1:]...)
	return nil
}

// Clear removes all interceptors.
func (ic *InterceptorChain) Clear() {
	ic.mu.Lock()
	defer ic.mu.Unlock()
	ic.interceptors = ic.interceptors[:0]
}

// Len returns the number of interceptors.
func (ic *InterceptorChain) Len() int {
	ic.mu.RLock()
	defer ic.mu.RUnlock()
	return len(ic.interceptors)
}

// Execute runs the handler through the full interceptor chain.
func (ic *InterceptorChain) Execute(ctx context.Context, sql string, args []interface{}, handler Handler) (*Result, error) {
	ic.mu.RLock()
	chain := make([]Interceptor, len(ic.interceptors))
	copy(chain, ic.interceptors)
	ic.mu.RUnlock()

	// Build the chain from inside out
	current := handler
	for i := len(chain) - 1; i >= 0; i-- {
		interceptor := chain[i]
		next := current
		current = func(ctx context.Context, sql string, args []interface{}) (*Result, error) {
			return interceptor(ctx, sql, args, next)
		}
	}
	return current(ctx, sql, args)
}

// ---------------------------------------------------------------------------
// Built-in Interceptors
// ---------------------------------------------------------------------------

// LoggingInterceptor logs query execution details.
func LoggingInterceptor(onLog func(sql string, args []interface{}, duration time.Duration, err error)) Interceptor {
	return func(ctx context.Context, sql string, args []interface{}, next Handler) (*Result, error) {
		start := time.Now()
		result, err := next(ctx, sql, args)
		duration := time.Since(start)
		if onLog != nil {
			onLog(sql, args, duration, err)
		}
		return result, err
	}
}

// RetryInterceptor adds retry logic to queries.
func RetryInterceptor(policy *RetryPolicy) Interceptor {
	return func(ctx context.Context, sql string, args []interface{}, next Handler) (*Result, error) {
		return ExecuteResult(ctx, policy, func() (*Result, error) {
			return next(ctx, sql, args)
		})
	}
}

// CircuitBreakerInterceptor adds circuit breaker protection.
func CircuitBreakerInterceptor(breaker *CircuitBreaker) Interceptor {
	return func(ctx context.Context, sql string, args []interface{}, next Handler) (*Result, error) {
		return CallResult(breaker, func() (*Result, error) {
			return next(ctx, sql, args)
		})
	}
}

// TimeoutInterceptor enforces a timeout on queries.
func TimeoutInterceptor(timeout time.Duration) Interceptor {
	return func(ctx context.Context, sql string, args []interface{}, next Handler) (*Result, error) {
		ctx, cancel := context.WithTimeout(ctx, timeout)
		defer cancel()
		return next(ctx, sql, args)
	}
}

// MetricsInterceptor collects query metrics.
func MetricsInterceptor(metrics *MetricsCollector) Interceptor {
	return func(ctx context.Context, sql string, args []interface{}, next Handler) (*Result, error) {
		start := time.Now()
		result, err := next(ctx, sql, args)
		metrics.RecordQuery(time.Since(start))
		if err != nil {
			metrics.RecordQueryFailure()
		}
		return result, err
	}
}

// ValidationInterceptor validates queries before execution.
func ValidationInterceptor(maxLength int) Interceptor {
	return func(ctx context.Context, sql string, args []interface{}, next Handler) (*Result, error) {
		if len(sql) > maxLength {
			return nil, NewValidationError(fmt.Sprintf("query exceeds maximum length of %d", maxLength))
		}
		return next(ctx, sql, args)
	}
}

// CachingInterceptor adds query result caching.
func CachingInterceptor(cache *Cache) Interceptor {
	return func(ctx context.Context, sql string, args []interface{}, next Handler) (*Result, error) {
		if !isSelectQuery(sql) {
			return next(ctx, sql, args)
		}
		key := cache.QueryKey(sql, args)
		if entry := cache.Get(key); entry != nil {
			return entry.Result, nil
		}
		result, err := next(ctx, sql, args)
		if err == nil {
			cache.Set(key, sql, args, result)
		}
		return result, err
	}
}

// InterceptedClient wraps a Client with an interceptor chain.
type InterceptedClient struct {
	*Client
	chain *InterceptorChain
}

// NewInterceptedClient creates a client with an interceptor chain.
func NewInterceptedClient(client *Client, chain *InterceptorChain) *InterceptedClient {
	return &InterceptedClient{
		Client: client,
		chain:  chain,
	}
}

// Query executes a query through the interceptor chain.
func (ic *InterceptedClient) Query(ctx context.Context, sql string, args ...interface{}) (*Result, error) {
	handler := func(ctx context.Context, sql string, args []interface{}) (*Result, error) {
		return ic.Client.Query(ctx, sql, args...)
	}
	return ic.chain.Execute(ctx, sql, args, handler)
}

// Exec executes a statement through the interceptor chain.
func (ic *InterceptedClient) Exec(ctx context.Context, sql string, args ...interface{}) (int64, error) {
	result, err := ic.Query(ctx, sql, args...)
	if err != nil {
		return 0, err
	}
	return int64(result.RowCount), nil
}
